File: 백준/Gold/5427. 불/불.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const move = [
  [0, 1],
  [0, -1],
  [1, 0],
  [-1, 0],
]

const findAnswer = (fires, start, graph) => {
  let answer = Number.MAX_SAFE_INTEGER
  const w = graph.length
  const h = graph[0].length

  // 1행
  // 마지막 행
  for (let i = 0; i < h; i += 1) {
    if (fires[0][i] > start[0][i]) {
      answer = Math.min(answer, start[0][i])
    }
    if (fires[w - 1][i] > start[w - 1][i]) {
      answer = Math.min(answer, start[w - 1][i])
    }
  }
  // 1열
  // 마지막 열
  for (let i = 0; i < w; i += 1) {
    if (fires[i][0] > start[i][0]) {
      answer = Math.min(answer, start[i][0])
    }
    if (fires[i][h - 1] > start[i][h - 1]) {
      answer = Math.min(answer, start[i][h - 1])
    }
  }
  return answer === Number.MAX_SAFE_INTEGER ? "IMPOSSIBLE" : answer
}

const bfs = (pos, graph, isFire = false) => {
  const queue = [...pos]
  const w = graph.length
  const h = graph[0].length

  let index = 0
  const visited = Array.from({ length: w }, () =>
    Array.from({ length: h }, () => Number.MAX_SAFE_INTEGER),
  )

  for (let [x, y] of pos) {
    visited[x][y] = 1
  }

  while (queue.length > index) {
    const [x, y] = queue[index++]
    for (let i = 0; i < 4; i += 1) {
      const nx = x + move[i][0]
      const ny = y + move[i][1]
      if (
        nx < 0 ||
        ny < 0 ||
        nx >= w ||
        ny >= h ||
        graph[nx][ny] === "#" ||
        visited[nx][ny] <= visited[x][y] + 1
      ) {
        continue
      }

      if (!isFire && graph[nx][ny] === "*") {
        continue
      }
      visited[nx][ny] = visited[x][y] + 1
      queue.push([nx, ny])
    }
  }

  return visited
}

const T = Number(input[0])
let index = 1
for (let tc = 1; tc <= T; tc += 1) {
  const [w, h] = input[index++].split(" ").map(Number)
  const graph = input.slice(index, index + h).map((row) => row.split(""))
  index += h

  const firesPositions = []
  const startPositions = []
  for (let i = 0; i < h; i += 1) {
    for (let j = 0; j < w; j += 1) {
      if (graph[i][j] === "*") {
        firesPositions.push([i, j])
      } else if (graph[i][j] === "@") {
        startPositions.push([i, j])
      }
    }
  }

  const firesVisited = bfs(firesPositions, graph, true)
  const startVisited = bfs(startPositions, graph)

  // 상근이는 무조건 불이 붙는데 걸린 시간보다 적게 걸려야함. 같거나 크면 탈출 불가
  console.log(findAnswer(firesVisited, startVisited, graph))
}

File: 백준/Gold/5427. 불/불.py
import sys
from collections import deque
input = sys.stdin.readline

tc = int(input().strip())
answer =[]


move = [[1,0],[-1,0],[0,1],[0,-1]]
def bfs(targetPosition,graph,h,w,isFire=False):
  queue = deque(targetPosition)
  distance = [[1e6 for _ in range(w)] for _ in range(h)]
  for [x,y] in targetPosition:
    distance[x][y] = 1

  while(len(queue)>0):
    [x,y] = queue.popleft()
    for i in range(4):
      nx =x+move[i][0]
      ny =y+move[i][1]
      if nx<0 or ny<0 or nx>=h or ny>=w or graph[nx][ny] == '#' or distance[nx][ny] != 1e6: continue
      if(not isFire and graph[nx][ny] == '*'): continue
      distance[nx][ny] = distance[x][y] +1
      queue.append([nx,ny])

  return distance


  

def findAnswer(fire,start,h,w):
  tmp = []

  # 1행 마지막행 확인
  for i in range(w):
    if fire[0][i] > start[0][i]:
      tmp.append(start[0][i])
    if fire[h-1][i] > start[h-1][i]:
      tmp.append(start[h-1][i])
  
  # 1열 마지막열 확인
  for i in range(h):
    if fire[i][0] > start[i][0]:
      tmp.append(start[i][0])
    if fire[i][w-1] > start[i][w-1]:
      tmp.append(start[i][w-1])
  if(len(tmp) >0):
    return min(tmp)
  else:
    return 'IMPOSSIBLE'




for t in range(tc):
  w,h = map(int,input().split())
  graph = [list(input().strip()) for _ in range(h)]

  startPosition = []
  firePositions = []

  for i in range(h):
    for j in range(w):
      if(graph[i][j] == '@'):
        startPosition.append([i,j])
      elif (graph[i][j] == '*'):
        firePositions.append([i,j])

  fireDist = bfs(firePositions,graph,h,w,True)
  startDist = bfs(startPosition,graph,h,w)
  result = findAnswer(fireDist,startDist,h,w)
  print(result)


